lang/cs/Org.Apache.REEF.Network/Elastic/Topology/Logical/Impl/FlatTopology.cs

```diff
@@ -0,0 +1,448 @@
+
+                returnMessages.Add(returnMessage);
+
+                if (_nodesWaitingToJoinTopology.Count > 0)
+                {
+                    Log.Log(Level.Info,
+                        "Tasks [{0}] are added to topology in iteration {1}",
+                        string.Join(",", _nodesWaitingToJoinTopology),
+                        _iteration);
+
+                    _availableDataPoints += _nodesWaitingToJoinTopology.Count;
+                    failureStateMachine.Value.AddDataPoints(_nodesWaitingToJoinTopology.Count, false);
+
+                    foreach (var node in _nodesWaitingToJoinTopology)
+                    {
+                        var id = Utils.GetTaskNum(node);
+                        _nodes[id].FailState = DataNodeState.Reachable;
+                    }
+
+                    _nodesWaitingToJoinTopology.Clear();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Action to trigger when the operator receives a notification that a new iteration is started.
+        /// </summary>
+        /// <param name="iteration">The new iteration number</param>
+        public void OnNewIteration(int iteration)
+        {
+            Log.Log(Level.Info,
+                "Flat Topology for Operator {0} in Iteration {1} is closed with {2} nodes",
+                OperatorId,
+                iteration - 1,
+                _availableDataPoints);
+            _iteration = iteration;
+            _totNumberofNodes += _availableDataPoints;
+
+            lock (_lock)
+            {
+                _nodesWaitingToJoinTopology = _nodesWaitingToJoinTopologyNextIteration;
+                _nodesWaitingToJoinTopologyNextIteration = new HashSet<string>();
+            }
+        }
+
+        /// <summary>
+        /// Reconfigure the topology in response to some event.
+        /// </summary>
+        /// <param name="taskId">The task id responsible for the topology change</param>
+        /// <param name="info">Some additional topology-specific information</param>
+        /// <param name="iteration">The optional iteration number in which the event occurred</param>
+        /// <returns>One or more messages for reconfiguring the Tasks</returns>
+        public IList<IElasticDriverMessage> Reconfigure(
+            string taskId,
+            Optional<string> info,
+            Optional<int> iteration)
+        {
+            if (taskId == _rootTaskId)
+            {
+                throw new NotImplementedException("Failure on master not supported yet.");
+            }
+
+            List<IElasticDriverMessage> messages = new List<IElasticDriverMessage>();
+
+            lock (_lock)
+            {
+                int iter;
+
+                if (info.IsPresent())
+                {
+                    iter = int.Parse(info.Value.Split(':')[0]);
+                }
+                else
+                {
+                    iter = iteration.Value;
+                }
```

Comment:
one-liner:
```C#
int iter = info.IsPresent() ? int.Parse(info.Value.Split(':')[0]) : iteration.Value;
```
(can be outside of the lock, by the way)

Reply:
Fixed.